uqcsbot/utils/advent_utils.py

```diff
@@ -366,21 +398,81 @@ def parse_leaderboard_column_string(s: str, bot: UQCSBot) -> List[LeaderboardCol
     return columns
 
 
+def render_leaderboard_to_text(leaderboard: Leaderboard) -> str:
+    return "".join(x if isinstance(x, str) else x.text for x in leaderboard)
+
+
+def _isolate_leaderboard_layers(
+    leaderboard: Leaderboard,
+) -> Tuple[str, Dict[Colour, str]]:
+    """
+    Given a leaderboard made up of coloured fragments, split the
+    text into a number of layers. Each layer contains all the text
+    which is coloured by one particular colour.
+
+    Returns:
+    - a string of the leaderboard, but with whitespace in the place of every character,
+      for calculating bounding box size.
+    - a dictionary mapping colours to the layer of that colour.
+    """
+    layers: DefaultDict[str | None, str] = defaultdict(lambda: layers[None])
+    layers[None] = ""
+
+    for frag in leaderboard:
+        colour, text = (
+            (FG_COLOUR, frag) if isinstance(frag, str) else (frag.colour, frag.text)
+        )
+        layers[colour] += text
+        for k in layers:
+            if k == colour:
+                continue
+            layers[k] += "".join(c if c.isspace() else " " for c in text)
+
+    spaces = layers[None]
+    del layers[None]
+    return spaces, layers  # type: ignore
+
+
+def render_leaderboard_to_image(leaderboard: Leaderboard) -> bytes:
+    spaces, layers = _isolate_leaderboard_layers(leaderboard)
+
+    font = PIL.ImageFont.truetype("./uqcsbot/static/NotoSansMono-Regular.ttf", 20)
```

Comment:
For what it's worth, the AoC website uses Source Code Pro for its text, if we want to match it more closely